main/compiling/Circuit.py

```diff
@@ -427,19 +424,48 @@ def _to_stim(
         self.measurer.reset_compilation()
         return full_circuit
 
-    def instruction_to_stim(self, instruction: Instruction, circuit: stim.Circuit):
-        """Adds an individual Instruction to a circuit
+    def split_instructions_according_to_gate(self, qubit_instructions: Dict[Qubit, List[Instruction]]):
+        """Splits the instructions into gates and measurements
+
+        Qubit instructions is a dictionary whose keys are Qubits and values are lists of Instructions.
+        This function generates a dictionary whose keys are a instruction and whose values are lists of qubits.
+        This is done to reduce the amount of times we need to call Stim's append function.
+
+        It also generates a list of measurements.
 
         Args:
-            instruction: The instruction to be translated
-            circuit: A stim circuit to add the instruction to.
-        """
-        if instruction.targets is not None:
-            targets = instruction.targets
-        else:
-            targets = [self.qubit_index(qubit) for qubit in instruction.qubits]
-        circuit.append(instruction.name, targets, instruction.params)
+            qubit_instructions: A dictionary whose keys are Qubits and values are lists of Instructions.
 
+        Returns:
+            A tuple of two dictionaries. The first dictionary has keys of Instructions and values of lists of qubits.
+            The second dictionary has keys of Instructions which are measurements and values of lists of qubits.        
+        """
+        measurements = []
+        qubits_by_instruction = {}
+        compiled_instruction = defaultdict(bool)
+        instructions = sorted(qubit_instructions.values(),
+                              key=lambda value: (value[0].name, self.qubit_index(value[0].qubits[0])))
+        for instruction in instructions:
+
+            first_instruction = instruction[0]
```

Comment:
Definitely need to be sure there's only ever one instruction if writing it like this, right?

To be on the safe side, why not write a for loop for each instruction in instruction? Or am I missing something?

Reply:
So in the Circuit class' `add_instruction` method, can see that multiple instructions are allowed on a qubit at a given tick under some conditions - e.g. multiple **noise** instructions are allowed on the same qubit at the same tick, for example if we're adding noise because we've just done a gate, and more noise because we're about to do a measurement. (It also says there we allow multiple MPP instructions on the same qubit at the same tick, but I'm no longer sure why I thought this was a good idea..?)

Reply:
Ah I suspect the MPP thing was because we make a copy of the circuit to figure out initial detectors, and this copy just uses MPP measurements to keep things simple?? Dumb design from me tbf.

Reply:
Note to self for you via me: make a test!

Reply:
test added in test_Circuit.py